src/D2L.CodeStyle.Analyzers/ApiUsage/Serialization/SerializerAttributeAnalyzer.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage.Serialization {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class SerializerAttributeAnalyzer : DiagnosticAnalyzer {
+
+		private static readonly SymbolDisplayFormat TypeDisplayFormat = new SymbolDisplayFormat(
+				memberOptions: SymbolDisplayMemberOptions.IncludeContainingType,
+				localOptions: SymbolDisplayLocalOptions.IncludeType,
+				typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces
+			);
+
+		private const string SerializerAttributeFullName = "D2L.LP.Serialization.SerializerAttribute";
+		private const string ITrySerializerFullName = "D2L.LP.Serialization.ITrySerializer";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.InvalidSerializerType
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterScopeBuilderAnalyzer );
+		}
+
+		private void RegisterScopeBuilderAnalyzer( CompilationStartAnalysisContext context ) {
+
+			Compilation comp = context.Compilation;
+			if( !comp.TryGetTypeByMetadataName( SerializerAttributeFullName, out INamedTypeSymbol serializerAttributeType ) ) {
+				return;
+			}
+			if( !comp.TryGetTypeByMetadataName( ITrySerializerFullName, out INamedTypeSymbol trySerializerType ) ) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(
+					c => AnalyzeAttributeSyntax(
+						c,
+						(AttributeSyntax)c.Node,
+						serializerAttributeType,
+						trySerializerType
+					),
+					SyntaxKind.Attribute
+				);
+		}
+
+		private void AnalyzeAttributeSyntax(
+				SyntaxNodeAnalysisContext context,
+				AttributeSyntax attributeSyntax,
+				INamedTypeSymbol serializerAttributeType,
+				INamedTypeSymbol trySerializerType
+			) {
+
+			SemanticModel model = context.SemanticModel;
+			if( !IsAttributeOfType( model, attributeSyntax, serializerAttributeType ) ) {
+				return;
+			}
+
+			AttributeArgumentSyntax typeArgumentSyntax = attributeSyntax.ArgumentList.Arguments[ 0 ];
+			if( !( typeArgumentSyntax.Expression is TypeOfExpressionSyntax typeofSyntax ) ) {
+
+				ReportInvalidSerializerType(
+						context,
+						typeArgumentSyntax,
+						messageArg: typeArgumentSyntax.Expression.ToString()
+					);
+				return;
+			}
+
+			ITypeSymbol serializerType = model.GetTypeInfo( typeofSyntax.Type ).Type;

```

Comment:
Potential NRE if the type doesn't exist in the model. Consider handling null here as well.

```suggestion
			ITypeSymbol serializerType = model.GetTypeInfo( typeofSyntax.Type )?.Type;
			if( serializerType is null ) {
				// compilation will fail
			    return;
			}
```

Reply:
Ah, you're right, my bad. Seems like it's the `Type` itself that would be null then, which would NRE below in `serializerType.AllInterfaces`. Adding a test case for it would be good.

https://github.com/dotnet/roslyn/blob/e704ca635bd6de70a0250e34c4567c7a28fa9f6d/src/Compilers/CSharp/Portable/Compilation/TypeInfo.cs#L13

Reply:
Good feedback regardless.  Got me thinking about more `null` possibilities